src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -0,0 +1,428 @@
+        return;
+    }
+
+    match kind.unpack() {
+        UnpackedKind::Type(ty) => {
+            // `T: 'outlived_region` for some type `T`
+            // But T could be a lot of things:
+            // e.g., if `T = &'b u32`, then `'b: 'outlived_region` is what we want to add
+            //
+            // or if within `struct Foo<U>` you had `T = Vec<U>`, then we would
+            // want to add `U: 'outlived_region`
+            for component in tcx.outlives_components(ty) {
+                match component {
+                    Component::Region(r) => {
+                        // This would arise from something like:
+                        //
+                        // ```
+                        // struct Foo<'a, 'b> {
+                        //    x:  &'a &'b u32
+                        // }
+                        // ```
+                        //
+                        // Here `outlived_region = 'a` and `kind = &'b
+                        // u32`.  Decomposing `&'b u32` into
+                        // components would yield `'b`, and we add the
+                        // where clause that `'b: 'a`.
+                        insert_outlives_predicate(
+                            tcx,
+                            r.into(),
+                            outlived_region,
+                            required_predicates,
+                        );
+                    }
+
+                    Component::Param(param_ty) => { // param_ty: ty::ParamTy
+                        // This would arise from something like:
+                        //
+                        // ```
+                        // struct Foo<'a, U> {
+                        //    x:  &'a Vec<U>
+                        // }
+                        // ```
+                        //
+                        // Here `outlived_region = 'a` and `kind =
+                        // Vec<U>`.  Decomposing `Vec<U>` into
+                        // components would yield `U`, and we add the
+                        // where clause that `U: 'a`.
+                        let ty: Ty<'tcx> = tcx.mk_param(param_ty.idx, param_ty.name);
+                        required_predicates.insert(ty::OutlivesPredicate(ty.into(), outlived_region));
+                    }
+
+                    Component::Projection(proj_ty) => {
+                        // This would arise from something like:
+                        //
+                        // ```
+                        // struct Foo<'a, T: Iterator> {
+                        //    x:  &'a <T as Iterator>::Item
+                        // }
+                        // ```
+                        //
+                        // Here we want to add an explicit `where <T as Iterator>::Item: 'a`.
+                        let ty: Ty<'tcx> = tcx.mk_projection(proj_ty.item_def_id, proj_ty.substs);
+                        required_predicates.insert(ty::OutlivesPredicate(ty.into(), outlived_region));
```

Comment:
@nikomatsakis this correctly does inference for projection types as verified by the [passing test](https://github.com/toidiu/rust/blob/e0b04d75bd35bd8c25b9e124761cea40e822a158/src/test/ui/feature-gate-infer-outlives-requirements/projections-pass.rs).

Given that not sure if we need [ty::TyProjection](https://github.com/rust-lang/rust/pull/45298/files#diff-a43200d3360f31959b9f923bc33ed1edR218)? Your comment suggests that we need to use explicit predicates but I couldn't come up with a failing test. Could you provide an example of a failing tests please.